desktop/src/gui/controller.rs

```diff
@@ -511,46 +509,139 @@ fn try_wgpu_backend(backend: wgpu::Backends) -> Option<wgpu::Instance> {
-        _ => Some(Family::Name("Noto Sans")),
-    } {
-        families.push(linux_font);
-    }
-    families.extend(
-        [
-            Family::Name("Source Han Sans"),   // linux
-            Family::Name("WenQuanYi Zen Hei"), // linux, old
-            Family::Name("Arial Unicode MS"),  // macos
-            Family::SansSerif,
-        ]
-        .iter(),
+    locale: unic_langid::LanguageIdentifier,
+) -> egui::FontDefinitions {
+    let mut fd: FontDefinitions = egui::FontDefinitions::default();
+
+    let lang = locale.language.as_str();
+    let is_ja = lang == "ja";
+    let is_ko = lang == "ko";
+    let is_zh = lang == "zh";
+    let is_sc = is_zh && locale.to_string().as_str() == "zh-CN";
+    let is_tc = is_zh && !is_sc;
+
+    let mut queries: PrioritizedQueries = Vec::new();
+
+    // The main font
+    queries.push((1, vec![Family::SansSerif]));
+
+    // Pan-CJK fonts
+    queries.push((
+        2,
+        vec![
+            Family::Name("Noto Sans CJK"),     // Open font
+            Family::Name("Source Han Sans"),   // Open font, same as Noto Sans CJK
+            Family::Name("WenQuanYi Zen Hei"), // Open font
+            Family::Name("Arial Unicode MS"),  // MacOS
+        ],
+    ));
+
+    // Korean
+    queries.push((
+        3 + if is_ko { 0 } else { 1 },
+        vec![
+            Family::Name("Noto Sans CJK KR"), // Open font
+            Family::Name("Malgun Gothic"),    // Windows
+        ],
+    ));
+
+    // Japanese
+    queries.push((
+        3 + if is_ja { 0 } else { 1 },
+        vec![
+            Family::Name("Noto Sans CJK JP"), // Open font
+            Family::Name("MS UI Gothic"),     // Windows
+        ],
+    ));
+
+    // Chinese Simplified
+    queries.push((
+        3 + if is_sc { 0 } else { 1 },
+        vec![
+            Family::Name("Noto Sans CJK SC"), // Open font
+            Family::Name("Microsoft YaHei"),  // Windows
+        ],
+    ));
+
+    // Chinese Traditional
+    queries.push((
+        3 + if is_tc { 0 } else { 1 },
+        vec![
+            Family::Name("Noto Sans CJK TC"),   // Open font
+            Family::Name("Microsoft JhengHei"), // Windows
+        ],
+    ));
+
+    // Hebrew
+    queries.push((
+        4,
+        vec![
+            Family::Name("Noto Sans Hebrew"), // Open font
+        ],
+    ));
+
+    // Arabic
+    queries.push((
+        5,
+        vec![
+            Family::Name("Noto Sans Arabic"), // Open font
+        ],
+    ));
+
+    register_family(
+        font_database,
+        &mut fd,
+        egui::FontFamily::Proportional,
+        queries,
     );
 
+    fd
+}
+
+type FamilyQuery<'a> = Vec<Family<'a>>;
+type PrioritizedQueries<'a> = Vec<(usize, FamilyQuery<'a>)>;
+
+fn register_family(
+    font_database: &Database,
+    fd: &mut FontDefinitions,
+    family: egui::FontFamily,
+    mut queries: PrioritizedQueries<'_>,
+) {
+    queries.sort_by_key(|(priority, _)| *priority);
+    for (_, query) in queries {
```

Comment:
Maybe I a missing something, but shouldn't we start with the highest priority?
https://docs.rs/fontdb/latest/fontdb/struct.Query.html#structfield.families
This references CSS font-family, where the first has the highest priority.

Reply:
Oh, I am being stupid. We already higher priority == lower number.

Reply:
We do. I'm not sure what you're referring to, but note we have two kinds of priorities. `PrioritizedQueries` describe queries by priority (so priority 1 is the main font, 2 is the pan-CJK font, etc.), and each (CSS-like) query has a list of font families, where the first font that matches is returned.